# Python Notebook Test for DIGIT 210 (April 2026)

## Instructions
Copy this test file to your personal GitHub repository python workspace. Modify the notebook to add markdown and code cells as directed. Review and consult code files from our class repo in textAnalysis-Hub/Class-Examples as you work on the test. The test is about specifically what we learned in this class, so material that was not part of our readings, discussion, or class experience will not be helpful. 

This test is an effort to elevate and evaluate *your own understanding* of the Python unit. It's okay to look up information but please work with our course materials first and foremost to be sure you working within the framework of our class. Using only AI to assist you on this test without reference or understanding of the course material will result in code that does not work and answers that do not fit our context and aren't given *in your own words*, which we ask for throughout this test. If you do not understand how to work with files in an activated Python environment, you will not be able to properly complete this test. 

Answer each question by creating a code or markdown block beneath it for your response, choosing the type of cell as appropriate. Being able to choose the correct kind of cell for your response is part of the exam. *Total points: 50*

**Part A.** *For this portion of the test, respond by adding markdown cells. Each response is worth a possible total of 3 points.*

1. Text tokenization is super important in natural language processing! What is *tokenization* as you understand it? 


Tokenization is when you can take, text, and phrases, into single words and characters so a program can analyze it one at a time rather than as a big incohiernt form that it doesnt understand.

2. Which Python libraries have we worked with in this class that tokenize text to *study and classify* its language for analysis?


NLTK and spaCy are the two libraires we used.

3. How does a generative language model tokenize text inputs differently than these Python libraries? 
Provide a specific example from the readings for this class to show and explain how this tokenization is different. Clearly identify the class reading that this example came from in your response for full credit.

LLM breakdown 1 reading showed howe GPT2 tokenizes the sentence "I like the longer-form posts on Substack" it turned 41 characters into just 11 tokens, a 75% compression. there was also a wierd thing that happend with "like" and " like" and they are the same but counted as different. so NLTK/spaCy split word and senetences based on grammer rules but LLMs convert text intio numbers using subword chuncks. 

4. One of our readings played with the partial sentence "I like oat milk in my . . . " in order to demonstrate something about how language models are functioning using logits. Review this reading and describe in your own words what logits have to do with the output, why they are called logits, and what we did to change the output. 

Logits are raw numbers that the model outputs for every possible token, this word comes from statistics its related to the logprithm of an odds ratio. The Oat milk example "i like oat milk in my..." the model would predict mouth as the next token becasue it had tyhe highest logit. To change it they would use a hook function to manually inflate the logit for "coffee" by 10 units.

5. What makes "France" a distractor in the LLM Breakdown 3 reading/notebook? Review this and explain how you understand what's going on with the token "France" in this notebook.

France is a distractor as it was never in the origonal text the sentance was the captial of germany is... the experiment swapped germany embeddings vector with france's using  a hook function which caused the model to predict paris instead of berlin. This showed that the model predictions depend on the embeddings vectors. 

6. How do you activate your virtual environment for running Python? Explain what you do, what command you enter, and how you execute it based on your local computer setup. 

navigate to my directiory i wanna use open either git bash or my termianl and run the command <source Scripts/activate> and see it appear.

7. If you don't set up a `.gitignore` on a GitHub repository, what are two serious problems you may encounter? How does the `.gitignore` prevent these problems? And why does it start with a dot `.` ?

two mian problems with be it wil add files to your repo that you dont want to be public and hold sensitive info, and it pushed your venv folder to your repo wich is big and uneeded. The .gitignore stops these issues by telling git to ignore the files and folders. It starts with a dot becasue dot files are hidden systems on unix based systems. 

**Part B.** *For this portion of the test, answer by preparing code cells and adding your own comments as Python comments. You may also include a markdown cell if helpful for your explanation* 

This is the "text lab" application part of the test.
* If you haven't done so already, prepare a virtual environment for this test that you activate to run Python 3.12.
* Prepare two directories of text files from your project and save them in the directory where you are working on this test.
You will need to demonstrate that you can write file paths to open and read and process these files, as well as to prepare output file(s).

1a. First, let's verify the details of your virtual environment. Run the following code, and **then paste in your output** in a markdown cell afterwards. (We're going to ask you to clean up your notebook and clear its output cells before you submit, so we need you to paste in the output in the following cell to preserve the info on your setup.) *(1 pts)*

In [1]:
import sys
print(sys.version)
print(sys.executable)

3.12.0 (tags/v3.12.0:0fb18b0, Oct  2 2023, 13:03:39) [MSC v.1935 64 bit (AMD64)]
i:\Github\Digit-210\Python Notebook Test\venv\Scripts\python.exe


**Click on this markdown cell and paste the results from executing the code block above here.**






3.12.0 (tags/v3.12.0:0fb18b0, Oct  2 2023, 13:03:39) [MSC v.1935 64 bit (AMD64)]
c:\Users\gwr41\AppData\Local\Programs\Python\Python312\python.exe

Your task is to prepare python code and install + import the necessary libraries for natural language processing to analyze your corpora of text files to do two things:
* Compare the two groups of text (or text corpora) for their **lexical diversity**, and
* Find out about which verbs or adjectives (your choice) are likely to have high **ambiguity**.

1b. **In the next cell,** indicate what we need to do to install in your python environment to reproduce your work. 
Be clear about how to do the installation, as if you were explaining this to a student just learning it for the first time.
Provide the installation lines and your explanation in the next markdown cell.  *(3 pts)*

To reproduce this work, we need to install Python libraries used for NLP These libraries can help us analyze text, such as measuring lexical diversity and identifying ambiguous words.
First, make sure you have a virtual environment activated i use Python 3.12 as that version works with lots of things unlike newer versons. Then install the required libraries using pip:
pip install nltk spacy
after that we need to download additional language data for these libraries.
NLTK run Python and enter:
import nltk
nltk.download('punkt')
nltk.download('averaged_perceptron_tagger')
nltk.download('wordnet')
These datasets allow us to tokenize text, tag parts of speech like verbs and adjectives, and analyze word meanings.
For spaCy, we install a language model with:
python -m spacy download en_core_web_sm
This model allows us to process English text and perform tasks like part-of-speech tagging and linguistic analysis.

2. Now, prepare a code block that includes the import lines needed for these tasks. (You may want to do this in stages and modify the blocks as you're working on the exam.) *(3 pts)*

In [2]:
import nltk
import spacy
from nltk.tokenize import word_tokenize
from nltk.corpus import wordnet
from nltk.probability import FreqDist
import os

In [3]:
nltk.download('punkt')
nltk.download('averaged_perceptron_tagger')
nltk.download('wordnet')
nltk.download('punkt_tab')
nltk.download('averaged_perceptron_tagger_eng')

[nltk_data] Downloading package punkt to
[nltk_data]     C:\Users\gwr41\AppData\Roaming\nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package averaged_perceptron_tagger to
[nltk_data]     C:\Users\gwr41\AppData\Roaming\nltk_data...
[nltk_data]   Package averaged_perceptron_tagger is already up-to-
[nltk_data]       date!
[nltk_data] Downloading package wordnet to
[nltk_data]     C:\Users\gwr41\AppData\Roaming\nltk_data...
[nltk_data]   Package wordnet is already up-to-date!
[nltk_data] Downloading package punkt_tab to
[nltk_data]     C:\Users\gwr41\AppData\Roaming\nltk_data...
[nltk_data]   Unzipping tokenizers\punkt_tab.zip.
[nltk_data] Downloading package averaged_perceptron_tagger_eng to
[nltk_data]     C:\Users\gwr41\AppData\Roaming\nltk_data...
[nltk_data]   Unzipping taggers\averaged_perceptron_tagger_eng.zip.


True

3a. Tell us about the texts/text collections you'll be comparing. Are they a portion of the text files you're working on in your project? We'd like you to include them in two folders with this test. Will we easily find these files in your repo? (Use the markdown cell below). *(2 pts)*

We are comparing episodes in jopjos bizarre adventure part 4.

3b. Prepare a code block that depends on running the previous code block and then reads from each directory of files and calculates what's needed to indicate **lexical diversity**.
    Running the cell that you prepare should reveal the lexical diversity of each set of files. *(3 pts)*

In [8]:
import os

folder1 = "part5"
folder2 = "part4"

def get_lexical_diversity(folder_path):
    all_tokens = []
    for filename in os.listdir(folder_path):
        if filename.endswith(".txt"):
            with open(os.path.join(folder_path, filename), "r", encoding="utf-8") as f:
                text = f.read()
                tokens = word_tokenize(text.lower())
                all_tokens.extend(tokens)
    unique_tokens = set(all_tokens)
    diversity = len(unique_tokens) / len(all_tokens)
    return diversity, len(all_tokens), len(unique_tokens)

div1, total1, unique1 = get_lexical_diversity(folder1)
div2, total2, unique2 = get_lexical_diversity(folder2)

print(f"Folder 1 ({folder1}): {total1} tokens, {unique1} unique, diversity = {div1:.4f}")
print(f"Folder 2 ({folder2}): {total2} tokens, {unique2} unique, diversity = {div2:.4f}")

Folder 1 (part5): 156422 tokens, 5364 unique, diversity = 0.0343
Folder 2 (part4): 148516 tokens, 6048 unique, diversity = 0.0407


4. Comment what the results are showing you about your texts in your own words. From what you can see, which set has the highest lexical diversity, and what does that mean about the language that set of files, compared to the other set? *(3 pts)* 

Folder 1 (part5): 156422 tokens, 5364 unique, diversity = 0.0343
Folder 2 (part4): 148516 tokens, 6048 unique, diversity = 0.0407
folder 2has slightly higher lexical diversity but Both are quite low overall.

5. Now, write a code block that depends on variables you set for your file directories in the previous code block(s). This codeblock should just select a set of tokens from your text based on their part of speech, either adjectives or verbs.  *(5 pts)* 

In [10]:
def get_verbs(folder_path):
    all_verbs = []
    for filename in os.listdir(folder_path):
        if filename.endswith(".txt"):
            with open(os.path.join(folder_path, filename), "r", encoding="utf-8") as f:
                text = f.read()
                tokens = word_tokenize(text.lower())
                tagged = nltk.pos_tag(tokens)
                verbs = [word for word, tag in tagged if tag.startswith("VB")]
                all_verbs.extend(verbs)
    return all_verbs

verbs1 = get_verbs(folder1)
verbs2 = get_verbs(folder2)

print(f"Folder 1 verb count: {len(verbs1)}")
print(f"Folder 2 verb count: {len(verbs2)}")

Folder 1 verb count: 24343
Folder 2 verb count: 23263


6. Develop the code in the previous block by copying it into the next code block, and here:
   * Develop code to find out about the top 15 most ambiguous words in your adjective or verb token list, and
   * Write your output to a neatly formatted text file for reading. The text output file should provide a clear heading to provide a title for each corpus you're comparing. *(5 pts)*

In [11]:
def get_ambiguous_words(verbs, top_n=15):
    freq = FreqDist(verbs)
    ambiguity = {}
    for word in freq.keys():
        synsets = wordnet.synsets(word, pos=wordnet.VERB)
        if synsets:
            ambiguity[word] = len(synsets)
    sorted_words = sorted(ambiguity.items(), key=lambda x: x[1], reverse=True)
    return sorted_words[:top_n]

ambiguous1 = get_ambiguous_words(verbs1)
ambiguous2 = get_ambiguous_words(verbs2)

# write to output file
with open("ambiguity_output.txt", "w", encoding="utf-8") as f:
    f.write("=== Corpus 1: Episodes (04) ===\n\n")
    for word, count in ambiguous1:
        f.write(f"{word}: {count} synsets\n")
    f.write("\n=== Corpus 2: Characters (03) ===\n\n")
    for word, count in ambiguous2:
        f.write(f"{word}: {count} synsets\n")

print("Output written to ambiguity_output.txt")

Output written to ambiguity_output.txt


7. What does it mean for these to be highly "ambiguous" words? Explain this in your own words based on what we learned about ambiguity from our readings and based on your project files. *(3 pts)*

In [ ]:
Highly ambiguous words can  be verbs that have many different meanings depending on the context as break has 59 sysets meaning it could have meant 59 different things such as bones breaking, walls, and more and with the conect of part 4 of jojos being called diamond is unbreakable it makes scenes.

All finished? As your last act worth points on this exam, clean up your notebook so that it's not displaying the outputs of your code executions. Tidy things up, push this to your personal repo, and send the "code view" link of this to Canvas (analogous to the link we sent you to start the exam). *(1 pt)*